src/transformStateWithClones.js

```diff
@@ -7,7 +7,45 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let addProperties = {};
+  let removeProperties = {};
+  let cleanProperties = {};
+  let addObject = state;
+
+  const array = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      addProperties = { ...addObject, ...action.extraData };
+      array.push(addProperties);
+      addObject = addProperties;
```

Comment:
After performing the action for each case, you can push the result to the 'array' once, outside of the if/switch blocks, to adhere to the 'Don't Repeat Yourself' principle.